src/search/pdbs/pattern_collection_generator_rcg.cc

```diff
@@ -0,0 +1,309 @@
+    /*
+      Start blacklisting after the percentage of total_max_time specified via
+      blacklisting_trigger_percentage has passed. Compute this time point once.
+    */
+    double blacklisting_start_time = total_max_time * blacklist_trigger_percentage;
+    while (true) {
+        check_blacklist_trigger_timer(blacklisting_start_time, timer);
+
+        unordered_set<int> blacklisted_variables =
+            get_blacklisted_variables(non_goal_variables);
+
+        int remaining_pdb_size_for_cegar = min(remaining_collection_size, max_pdb_size);
+        double remaining_time_for_cegar =
+            min(static_cast<double>(timer.get_remaining_time()), cegar_max_time);
+        Pattern pattern = generate_pattern_rcg(
+            remaining_pdb_size_for_cegar,
+            remaining_time_for_cegar,
+            goals[goal_index].var,
+            cegar_verbosity,
+            rng,
+            task_proxy,
+            cg_neighbors
+        );
+        handle_generated_pattern(
+            task_proxy,
+            pattern,
+            generated_patterns,
+            timer);
+
+        if (collection_size_limit_reached() ||
+            time_limit_reached(timer) ||
+            check_for_stagnation(timer)) {
+            break;
+        }
+
+        ++num_iterations;
+        ++goal_index;
+        goal_index = goal_index % goals.size();
+        assert(utils::in_bounds(goal_index, goals));
+    }
+
+    PatternCollectionInformation result =
+        get_pattern_collection(task_proxy, generated_patterns);
+    if (verbosity >= utils::Verbosity::NORMAL) {
+        utils::g_log << "Multiple RCG number of iterations: "
+                     << num_iterations << endl;
+        utils::g_log << "Multiple RCG average time per generator: "
+                     << timer.get_elapsed_time() / num_iterations
+                     << endl;
+        dump_pattern_collection_generation_statistics(
+            "Multiple RCG",
+            timer.get_elapsed_time(),
+            result);
+    }
+    return result;
+}
+
+static shared_ptr<PatternCollectionGenerator> _parse(options::OptionParser &parser) {
+    parser.add_option<int>(
+        "max_collection_size",
+        "maximum number of states in the pattern collection (ignored for the "
+        "initial collection consisting of singleton patterns for each goal "
+        "variable)",
+        "20000000",
+        Bounds("1", "infinity"));
+    parser.add_option<double>(
+        "total_max_time",
+        "maximum time in seconds for the multiple RCG algorithm. The "
+        "algorithm will always execute at least one iteration, i.e., call the "
+        "RCG algorithm once. This limit possibly overrides the limit "
+        "specified for the RCG algorithm.",
+        "100.0",
+        Bounds("0.0", "infinity"));
+    parser.add_option<double>(
+        "stagnation_limit",
+        "maximum time in seconds the multiple RCG algorithm allows without "
```

Comment:
Are the terms "RCG algorithm" and "multiple RCG algorithms" describing different things? This is really hard to understand. I think what is missing here is a synopsis entry that explains what the pattern generator actually does.